virtualmachine/gcp/util.go

```diff
@@ -435,3 +442,284 @@ func (svc *googleService) insertSSHKey() error {
+	return svc.service.Firewalls.Get(svc.vm.Project, svc.vm.Firewall).Do()
+}
+
+// addFirewallPorts adds new ports to a firewall
+func (svc *googleService) addFirewallPorts() error {
+
+	currFirewall, err := svc.getFirewall()
+	if err != nil {
+		return err
+	}
+	newRules := currFirewall.Allowed
+	for _, endpoint := range svc.vm.Endpoints {
+		newRules = append(newRules, &googlecloud.FirewallAllowed{
+			IPProtocol: endpoint.Protocol,
+			Ports:      endpoint.Ports,
+		})
+	}
+
+	return svc.patchFirewall(newRules)
+}
+
+// patchFirewall patches the firewall with the given allowed ports (rules)
+func (svc *googleService) patchFirewall(allowed []*googlecloud.FirewallAllowed) error {
+	newFirewall := &googlecloud.Firewall{
+		Allowed: allowed,
+	}
+	op, err := svc.service.Firewalls.Patch(svc.vm.Project, svc.vm.Firewall,
+		newFirewall).Do()
+	if err != nil {
+		return err
+	}
+	return svc.waitForGlobalOperationReady(op.Name)
+}
+
+// removeFirewallPorts removes ports from a given firewall
+func (svc *googleService) removeFirewallPorts() error {
+	firewall, err := svc.getFirewall()
+	if err != nil {
+		return err
+	}
+	// Initialize new rules with the current one
+	// Refer to this link for structure of firewall resource
+	// https://cloud.google.com/compute/docs/reference/latest/firewalls
+	newRules := firewall.Allowed
+	// Define a map to identify ports to be removed
+	remPorts := make(map[string]map[string]bool)
+	// Define a map to deal with elements where the same protocol
+	// appears again
+	encountered := make(map[string]bool)
+	// Identify ports for each protocol to be removed
+	for _, endpoint := range svc.vm.Endpoints {
+		// Initialize the map only if the given protocol is encountered
+		// for the first time
+		if !encountered[endpoint.Protocol] {
+			remPorts[endpoint.Protocol] = make(map[string]bool)
+			encountered[endpoint.Protocol] = true
+		}
+		for _, port := range endpoint.Ports {
+			remPorts[endpoint.Protocol][port] = true
+		}
+	}
+
+	// Remove the the ports and protocols that are to be removed from the
+	// rules. Iterate through each item in the list of rules.
+	for indEndp := 0; indEndp < len(newRules); indEndp++ {
+		allowed := newRules[indEndp]
+		// Iterate through each port in the port list
+		for indPort := 0; indPort < len(newRules[indEndp].Ports); indPort++ {
+			port := newRules[indEndp].Ports[indPort]
+			// Process removal if the given port is to be removed
+			if remPorts[allowed.IPProtocol][port] {
+				newRules[indEndp].Ports[indPort] = ""
+				newRules[indEndp].Ports = append(
+					newRules[indEndp].Ports[:indPort],
+					newRules[indEndp].Ports[indPort+1:]...
+				)
+				// Set the index back by one because we have
+				// shifted elements from indP onwards to left
+				// by 1
+				indPort = indPort - 1
+			}
+		}
+		// At the end of removal of ports for a particular protocol if
+		// there are no ports left, then remove that protocol as well.
+		if len(newRules[indEndp].Ports) == 0 {
```

Comment:
It is assumed that `Endpoint: <protocol,port>`
[NTH] If so, please comment that explicitly. or/and have consistency in code/comment by using either protocol or indEndp exclusively.

Reply:
The endpoints are defined as
```
[
  {
    "protocol": string,
    "ports": [
      string
    ]
  }
]
```
Each protocol can have multiple ports under it. for example:
```
[
  {
    "protocol": "tcp",
    "ports": [
      "8080",
      "9090"
    ]
  },
  {
    "protocol": "udp",
    "ports": [
      "123"
    ]
  }
]
```
What the comment means is - if there is no port under a protocol, then remove that protocol as well. So if udp:123 is to be removed, then after port 123 is removed, remove protocol "udp" too from the data structure. Hope this helps.